lib/network_manager.go

```diff
@@ -0,0 +1,787 @@
+	// de-indexing the validator would result in an excess number of outbound/inbound connections. Any excess connections
+	// will be cleaned up by the NonValidator connector.
+	validatorRemoteNodeMap := nm.rnManager.GetValidatorIndex().ToMap()
+	for pk, rn := range validatorRemoteNodeMap {
+		// If the validator is no longer active, de-index it.
+		if _, ok := activeValidatorsMap.Get(pk); !ok {
+			nm.rnManager.SetNonValidator(rn)
+			nm.rnManager.UnsetValidator(rn)
+		}
+	}
+
+	// Look for validators in our existing outbound / inbound connections.
+	allNonValidators := nm.rnManager.GetAllNonValidators()
+	for _, rn := range allNonValidators {
+		// It is possible for a RemoteNode to be in the non-validator indices, and still have a public key. This can happen
+		// if the RemoteNode advertised support for the SFValidator service flag during handshake, and provided us
+		// with a public key, and a corresponding proof of possession signature.
+		pk := rn.GetValidatorPublicKey()
+		if pk == nil {
+			continue
+		}
+		// It is possible that through unlikely concurrence, and malevolence, two non-validators happen to have the same
+		// public key, which goes undetected during handshake. To prevent this from affecting the indexing of the validator
+		// set, we check that the non-validator's public key is not already present in the validator index.
+		if _, ok := nm.rnManager.GetValidatorIndex().Get(pk.Serialize()); ok {
+			glog.V(2).Infof("NetworkManager.refreshValidatorIndex: Disconnecting Validator RemoteNode "+
+				"(%v) has validator public key (%v) that is already present in validator index", rn, pk)
+			nm.rnManager.Disconnect(rn)
+			continue
+		}
+
+		// If the RemoteNode turns out to be in the validator set, index it.
+		if _, ok := activeValidatorsMap.Get(pk.Serialize()); ok {
+			nm.rnManager.SetValidator(rn)
+			nm.rnManager.UnsetNonValidator(rn)
+		}
+	}
+}
+
+// connectValidators attempts to connect to all active validators that are not already connected. It is called
+// periodically by the validator connector.
+func (nm *NetworkManager) connectValidators(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]) {
+	// Look through the active validators and connect to any that we're not already connected to.
+	if nm.blsKeystore == nil {
+		return
+	}
+
+	validators := activeValidatorsMap.ToMap()
+	for pk, validator := range validators {
+		_, exists := nm.rnManager.GetValidatorIndex().Get(pk)
+		// If we're already connected to the validator, continue.
+		if exists {
+			continue
+		}
+		// If the validator is our node, continue.
+		if nm.blsKeystore.GetSigner().GetPublicKey().Serialize() == pk {
+			continue
+		}
+
+		publicKey, err := pk.Deserialize()
+		if err != nil {
+			continue
+		}
+
+		// For now, we only dial the first domain in the validator's domain list.
+		if len(validator.GetDomains()) == 0 {
+			continue
+		}
+		address := string(validator.GetDomains()[0])
```

Comment:
Should probably randomize across their domains or else everyone will pick the same one.